lib/edge_impulse/ei_ncs.cpp

```diff
@@ -0,0 +1,404 @@
+	}
+
+	if (looped) {
+		size_t buf_left = buf_end - read_start;
+
+		memcpy(b_res, read_start, buf_left * sizeof(float));
+		memcpy(b_res + buf_left, buf_start,
+		       (len - buf_left) * sizeof(float));
+	} else {
+		memcpy(b_res, read_start, len * sizeof(float));
+	}
+}
+
+static int buf_processing_move(struct data_buffer *b, size_t move)
+{
+	float *buf_start = b->buf;
+	float *buf_end = buf_start + ARRAY_SIZE(b->buf);
+	int res = 0;
+
+	k_spinlock_key_t key = k_spin_lock(&b->lock);
+
+	if (b->state == STATE_READY) {
+		b->state = STATE_PROCESSING;
+	} else {
+		__ASSERT_NO_MSG(b->state != STATE_DISABLED);
+		k_spin_unlock(&b->lock, key);
+		return -EBUSY;
+	}
+
+	size_t max_move = buf_calc_max_processing_move(b);
+
+	b->processing_head += move;
+	if (b->processing_head >= buf_end) {
+		b->processing_head -= ARRAY_SIZE(b->buf);
+	}
+
+	size_t processing_end_move = move + INPUT_WINDOW_SIZE;
+
+	if (processing_end_move > max_move) {
+		b->wait_data_size = processing_end_move - max_move;
+	} else {
+		res = PROCESS_BUF;
+	}
+
+	k_spin_unlock(&b->lock, key);
+
+	return res;
+}
+
+size_t ei_ncs_get_frame_size(void)
+{
+	return INPUT_FRAME_SIZE;
+}
+
+size_t ei_ncs_get_window_size(void)
+{
+	return INPUT_WINDOW_SIZE;
+}
+
+int ei_ncs_add_data(const float *data, const size_t data_size)
+{
+	if (data_size % INPUT_FRAME_SIZE) {
+		return -EINVAL;
+	}
+
+	int res = buf_append(&ei_input, data, data_size);
+
+	if (res == PROCESS_BUF) {
+		k_sem_give(&ei_sem);
+		res = 0;
+	}
+
+	return res;
+}
+
+int ei_ncs_clear_data(void)
+{
+	return buf_cleanup(&ei_input);
+}
+
+int ei_ncs_start_prediction(const size_t window_shift, const size_t frame_shift)
+{
+	size_t sample_shift = window_shift * ei_ncs_get_window_size() +
+			      frame_shift * ei_ncs_get_frame_size();
+
+	int ret = buf_processing_move(&ei_input, sample_shift);
+
+	if (ret == PROCESS_BUF) {
+		k_sem_give(&ei_sem);
+		ret = 0;
+	}
+
+	return ret;
+}
+
+static int raw_feature_get_data(size_t offset, size_t length, float *out_ptr)
+{
+	buf_get(&ei_input, out_ptr, offset, length);
+
+	return 0;
+}
+
+static void processing_finished(int err)
+{
+	__ASSERT_NO_MSG(user_cb);
+
+	buf_processing_end(&ei_input);
+	user_cb(err);
+}
+
+static void edge_impulse_thread_fn(void)
+{
+	signal_t features_signal;
+	features_signal.total_length = INPUT_WINDOW_SIZE;
+	features_signal.get_data = &raw_feature_get_data;
+
+	while (true) {
+		k_sem_take(&ei_sem, K_FOREVER);
+
+		/* Invoke the impulse. */
+		EI_IMPULSE_ERROR err = run_classifier(&features_signal,
+						      &ei_result, DEBUG_MODE);
+
+		if (err) {
+			LOG_ERR("run_classifier err=%d", err);
+		}
+
+		processing_finished(err);
+	}
+}
+
+static bool can_read_result(void)
+{
+	/* User is allowed to access results only from the result ready callback. */
+	return (k_current_get() == ei_thread_id);
+}
+
+int ei_ncs_get_classification_results(const char **label, float *value,
+				      float *anomaly)
+{
+	if (!can_read_result()) {
+		return -ENODATA;
```

Comment:
great check but this seems to access violation no lack of memory. `-EACCESS`?
plus add some warning message

Reply:
check other similar places

Reply:
done